File: dongle/matcher.py
import re
import heapq

_SEP = re.compile(r"[ /\\]+")
_DISPLAY_LIMIT = 50   # keep top N candidates; we show at most 15


def _score(q: str, q_segs_wrapped: list, path_lower: str, path_len: int) -> int:
    """
    Score a single path against a pre-processed query.

    q              : lowercased full query string
    q_segs_wrapped : query segments pre-wrapped with '/' for fast substring checks
    path_lower     : already-lowercased path
    path_len       : len(original path)
    """
    if q == path_lower:
        return 1_000_000

    score = 0

    # Substring match — most common fast path, checked before segment split
    idx = path_lower.find(q)
    if idx >= 0:
        # Segment-exact bonus on top of substring score
        # Optimized: Wrapping the path and segments in '/' avoids overhead from split()
        if q_segs_wrapped:
            path_wrapped = f"/{path_lower}/"
            for qsw in q_segs_wrapped:
                if qsw in path_wrapped:
                    score += 50_000
        score += 10_000 - idx
        return score

    # Segment-exact bonus (no full substring match)
    if q_segs_wrapped:
        path_wrapped = f"/{path_lower}/"
        for qsw in q_segs_wrapped:
            if qsw in path_wrapped:
                score += 50_000

    # If we already have segment bonus points, return without fuzzy
    if score > 0:
        return score

    # Fuzzy character-sequence match (most expensive, last resort)
    # Optimized: Using C-optimized `str.find` instead of a manual Python loop
    # significantly improves performance on fallback fuzzy matches.
    idx = -1
    for c in q:
        idx = path_lower.find(c, idx + 1)
        if idx == -1:
            return 0

    return 1_000 - path_len


def search(query: str, paths: list, frecency: dict = None) -> list:
    """
    Filter and rank paths against query.

    - Pre-computes query segments once (not per path).
    - Applies frecency boost when provided.
    - Returns paths sorted best-first.
    """
    if not paths:
        return []

    # Optimized: Hoisting the isinstance check outside the loop avoids evaluating
    # it thousands of times per keystroke during filtering and sorting.
    is_tuple = isinstance(paths[0], (tuple, list))

    # Pre-fetch dictionary method to reduce lookup overhead in hot loops
    fget = frecency.get if frecency else None

    if not query:
        if frecency:
            if is_tuple:
                # Optimized: native tuple comparison
                # Note: `i` is used as a tie-breaker so it doesn't fall back to comparing unorderable `p` objects
                # Optimized: using list comprehension instead of generator expression inside sorted() avoids generator overhead
                return [p for _, _, _, p in sorted([(-fget(p[1], 0), len(p[0]), i, p) for i, p in enumerate(paths)])]
            else:
                # Optimized: using list comprehension instead of generator expression inside sorted() avoids generator overhead
                return [p for _, _, _, p in sorted([(-fget(p, 0), len(p), i, p) for i, p in enumerate(paths)])]
        else:
            if is_tuple:
                return sorted(paths, key=lambda p: len(p[0]))
            else:
                return sorted(paths, key=len)

    q = query.lower()
    q_segs = [s for s in _SEP.split(q) if s]
    # PRE-COMPUTE: wrapping query segments in '/' to allow fast substring check
    q_segs_wrapped = [f"/{s}/" for s in q_segs]

    scored = []
    if is_tuple:
        for i, p in enumerate(paths):
            display = p[0]
            display_len = len(display)
            s = _score(q, q_segs_wrapped, display.lower(), display_len)
            if s > 0:
                if fget:
                    s += fget(p[1], 0) * 10
                # Optimized: structure tuple to leverage native element-by-element comparison
                # Note: `-i` is used as a tie-breaker so it preserves stable sorting order on `reverse=True`
                scored.append((s, -display_len, -i, p))
    else:
        for i, p in enumerate(paths):
            p_len = len(p)
            s = _score(q, q_segs_wrapped, p.lower(), p_len)
            if s > 0:
                if fget:
                    s += fget(p, 0) * 10
                # Optimized: structure tuple to leverage native element-by-element comparison
                scored.append((s, -p_len, -i, p))

    if not scored:
        return []

    # Partial sort: only need the top _DISPLAY_LIMIT entries
    # Optimized: native tuple sorting without lambda overhead
    if len(scored) > _DISPLAY_LIMIT:
        top = heapq.nlargest(_DISPLAY_LIMIT, scored)
        top.sort(reverse=True)
        return [p for _, _, _, p in top]

    scored.sort(reverse=True)
    return [p for _, _, _, p in scored]

File: lib/src/matcher.dart
const int _displayLimit = 50;

class MatchResult implements Comparable<MatchResult> {
  final int score;
  final int length;
  final int index;
  final dynamic path;

  MatchResult(this.score, this.length, this.index, this.path);

  @override
  int compareTo(MatchResult other) {
    if (score != other.score) return other.score.compareTo(score);
    if (length != other.length) return length.compareTo(other.length);
    return index.compareTo(other.index);
  }
}

int _score(String q, List<String> qSegsWrapped, String pathLower, int pathLen) {
  if (q == pathLower) return 1000000;

  int score = 0;

  int idx = pathLower.indexOf(q);
  if (idx >= 0) {
    if (qSegsWrapped.isNotEmpty) {
      String pathWrapped = '/$pathLower/';
      for (final qsw in qSegsWrapped) {
        if (pathWrapped.contains(qsw)) {
          score += 50000;
        }
      }
    }
    score += 10000 - idx;
    return score;
  }

  if (qSegsWrapped.isNotEmpty) {
    String pathWrapped = '/$pathLower/';
    for (final qsw in qSegsWrapped) {
      if (pathWrapped.contains(qsw)) {
        score += 50000;
      }
    }
  }

  if (score > 0) return score;

  idx = -1;
  for (int i = 0; i < q.length; i++) {
    idx = pathLower.indexOf(q[i], idx + 1);
    if (idx == -1) return 0;
  }

  return 1000 - pathLen;
}

List<dynamic> search(
  String query,
  List<dynamic> paths, {
  Map<String, int>? frecency,
}) {
  if (paths.isEmpty) return [];

  bool isTuple = paths[0] is List;

  if (query.isEmpty) {
    if (frecency != null && frecency.isNotEmpty) {
      var scored = <MatchResult>[];
      for (int i = 0; i < paths.length; i++) {
        var p = paths[i];
        if (isTuple) {
          int score = frecency[p[1]] ?? 0;
          scored.add(MatchResult(score, (p[0] as String).length, i, p));
        } else {
          int score = frecency[p] ?? 0;
          scored.add(MatchResult(score, (p as String).length, i, p));
        }
      }
      scored.sort();
      return scored.map((e) => e.path).toList();
    } else {
      var sortedPaths = List<dynamic>.from(paths);
      if (isTuple) {
        sortedPaths.sort(
          (a, b) => (a[0] as String).length.compareTo((b[0] as String).length),
        );
      } else {
        sortedPaths.sort(
          (a, b) => (a as String).length.compareTo((b as String).length),
        );
      }
      return sortedPaths;
    }
  }

  String q = query.toLowerCase();
  final sepRegex = RegExp(r'[ /\\]+');
  List<String> qSegs = q.split(sepRegex).where((s) => s.isNotEmpty).toList();
  List<String> qSegsWrapped = qSegs.map((s) => '/$s/').toList();

  var scored = <MatchResult>[];

  if (isTuple) {
    for (int i = 0; i < paths.length; i++) {
      var p = paths[i];
      String display = p[0] as String;
      int displayLen = display.length;
      int s = _score(q, qSegsWrapped, display.toLowerCase(), displayLen);
      if (s > 0) {
        if (frecency != null && frecency.isNotEmpty) {
          s += (frecency[p[1]] ?? 0) * 10;
        }
        scored.add(MatchResult(s, displayLen, i, p));
      }
    }
  } else {
    for (int i = 0; i < paths.length; i++) {
      var p = paths[i];
      String pStr = p as String;
      int pLen = pStr.length;
      int s = _score(q, qSegsWrapped, pStr.toLowerCase(), pLen);
      if (s > 0) {
        if (frecency != null && frecency.isNotEmpty) {
          s += (frecency[pStr] ?? 0) * 10;
        }
        scored.add(MatchResult(s, pLen, i, pStr));
      }
    }
  }

  if (scored.isEmpty) return [];

  scored.sort();

  if (scored.length > _displayLimit) {
    return scored.take(_displayLimit).map((e) => e.path).toList();
  }

  return scored.map((e) => e.path).toList();
}
